### ISSUE TITLE
Use RaspAP/lighttpd as the Captive Portal splash page

### ISSUE THREAD

**Issue by basconesherman** [openNDS/openNDS#266]
Hello. Im really new on Raspberry Pi, also on network things related to this topic. I have a web app, it has a login functionality. There are admin and guest users. All you need is to enter a password and the system will automatically detect if its an admin or a guest.

The admin will have functionalities to manage users. It can add another admin but in the guest tab, it will generate random passwords and tell the app about it. The admin can also change images and background in the login page. All things are managed using .txt files and PHP.

All things about the App is working fine so theres nothing about it. 

All I want is, use openNDS to broadcast this app. And when the password entered is a guest, it will grant internet connection. When its an admin, it will go directly on the app's system. 

Im using RaspAP alongside openNDS

All things are working on the openNDS but I want to point Captive Portal to my app which is hosted using lighttpd.

I hope you can help me. Im stuck on this for 2 weeks now. Thanks for your work btw, its really cool!

**Comment by basconesherman** [comment]
This my setup on this one

RaspAP's address is 10.3.141.1
openNDS gateway address is 10.3.141.1

When Im connected and browsed thru 10.3.141.1 on my browser, I can see the RaspAP GUI

So basically all things are working fine. I just replaced the RaspGUI files and copied my WebApp's file, so when I go to 10.3.141.1 I can see my app. But im doing it manually. I want to automate it using openNDS and when a guest user is granted, I want openNDS to grant the guest internet connection. Thank you. I hope you'll see this

**Comment by bluewavenet** [comment]
@basconesherman 
What you need is to add a FAS service to your lighttpd.
I suggest you read in detail all the documentation at https://opennds.readthedocs.io

Then study in in depth the FAS section:
https://opennds.readthedocs.io/en/stable/fas.html

If you are still stuck after this, then ask.

Hint: There are example FAS scripts included that work unmodified on lighttpd. All you will need is to integrate into your webapp for the required result.

PS. The RPi is a very poor choice for use as a router, especially for someone who is "really new on Raspberry Pi, also on network things". The wireless chip is closed source and the open source hack to make it work is very limited and intended only for use in STA mode. In AP mode the chipset will work but performance is poor and with more than 2 or 3 clients will be unreliable. For anything more than an academic exercise, you will be forced to use vlans, a managed switch and external wireless access points to make it work properly.

**Comment by bluewavenet** [comment]
@basconesherman 
Keep re reading the documentation!  It is all in there ;)
What version of opennds are you using?
Lightty cannot use port 80 as it is used by opennds. So you must chage the port lightty is using if you have not already.
This port will be fasport.
Remoteip is the ip address of lightty.

**Comment by basconesherman** [comment]
Ohh okay. I will keep on reading your documentation and some of the other closed issues here. Thanks for your time. It must be frustrating for you answering this dumb questions. Hehe. 

Well, I will. Close this and will just reopen if Im stuck again. Thanks!

**Comment by bluewavenet** [comment]
@basconesherman 

> It must be frustrating for you answering this dumb questions. 

They are not dumb questions. Asking questions is the best way of learning about a subject. I would prefer to encourage people to read the docs first though!
Documentation is not always good, so questions raised after reading are always useful for improving the docs.
So don't worry about asking ;-)

**Comment by basconesherman** [comment]
Anyways, I able to load my lighttpd server on openNDS. I just created a sample php page with the echo im here. Its loading and displaying when I clicked sign-in on android.

But when I use a link, for example css and javascript (jquery.min),  there is an error. I dont they are loaded on the page. 

I even tried displaying image but no luck. 
Do i need to have special functions to load external files on fas?

**Comment by basconesherman** [comment]
In addition, the css and js files are within the folder of lighttpd, so it dont need internet connection to load it. 

Also, im reading files on a .txt within the portal also, and its loading fine. I dont know why js and css is not loaded but .txt files are loaded.

If you'll ask, I have set of settings for my app located on .txt file, and the php is loading those data. Example, i got the link of the image from the textfile. 

The url is loading so I think the app is working just fine. But the image is not loading even though the link is ok. 

When I close openNDS, and just loading the sever from lighttpd, all things are working. So i know that there is no wrong in the app. Im just thinking that its on openNDS and ai have something whic i didnt set i suppose? I cant really answer this question from the docs, im sorry :(

**Comment by bluewavenet** [comment]
@basconesherman 

> I cant really answer this question from the docs

Well, actually:
https://opennds.readthedocs.io/en/stable/customize.html#rules-for-customised-splash-pages

However it is easy to miss (if you are reading without paying attention LOL).

The CPD popup browser is restricted for security reasons. Imagine, for example, if the captive portal was set up for a "Man In The Middle" attack. A downloaded js program can make a very good "Trojan" payload.

**Comment by basconesherman** [comment]
Ohhh. Yes youre right, I missed that one! 

So to be clear, css files can be loaded within the file. And inline css, but there is no way that a javascript can be loaded even the script is not external? Even the image?

**Comment by bluewavenet** [comment]
Local images are ok, local css is ok. html5 built in js is ok. External images and external content would require a Walled Garden setup in the openNDS config.

**Comment by basconesherman** [comment]
In my case, i copied my portal folder to the lighttpd path ( /var/www/html/portal)

All things are there like images and css. 

So on my index page, I write
<img src="img/bg.png" />
(portal/img) - its in the portal

But I have no luck loading it. Is this the part where I need to understand the Walled Garden you are saying?





But when

**Comment by bluewavenet** [comment]
Connect with a laptop/desktop machine. View the source of the code your FAS sent (ctrl u in Firefox for example)
You will be able to see where it is pointing for the image.. and what is wrong.

**Comment by basconesherman** [comment]
Yes that is the weird part. I can see no error on it. Anyways, thanks! This open a lot of workarounds that I can make to my app following those rules of captive portal. I thing I can handle these errors now since the networking part is already established.

**Comment by bluewavenet** [comment]
Also you can try copying fas-hid.php or fas-aes.php (depending on your config). Leave them unchanged and see if it works.
With the correct config these two examples work fine.

**Comment by basconesherman** [comment]
Hello I just want to inform you that I made it! Though I have to change almost half of my PHP App, I able to do a test and mock of all the functions, except for the upload feature. So I will have to tell my client that there are some limitations.

I loaded your sample fas-aes.php codes and inspect it on the browser and I saw my mistake why the images and css are not loading.

I must point the source to the openNDS path. And copy all the css and images to it. Inserted the openNDS address and voila, all things are loaded. :)

Also, the href limitations are still an issue. But I noticed that in your sample, there are many functions which you encapsulate the pages, thank you page, initial form and so on. So I usee the form action/get to imitate this and all are wroking just fine. 

Thank you for you patience, Ive learned a lot by using your works. Be safe brother! :)
